src/mate/academy/hashMap/ReferenceMyHashMap.java

```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Override
+    public void clear() {
+        Node<K, V>[] tab;
+        threshold++;
+        if ((tab = hashTable) != null && size > 0) {
+            size = 0;
+            for (int i = 0; i < tab.length; ++i)
+                tab[i] = null;
+        }
+    }
+
+    @Override
+    public boolean remove(final K key) {
+        int index = hash(key);
+        if (hashTable[index] == null)
+            return false;
+        if (hashTable[index].getNodes().size() == 1) {
+            hashTable[index] = null;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public V get(final K key) {
+        int index = hash(key);
+        if (index < hashTable.length &&
+                hashTable[index] != null) {
+            if (hashTable[index].getNodes().size() == 1) {
+                return hashTable[index].getNodes().get(0).getValue();
+            }
+
+            List<Node<K, V>> list = hashTable[index].getNodes();
+            for (Node<K, V> node : list) {
+                if (key.equals(node.getKey())) {
+                    return node.getValue();
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    private int hash(final K key) {
+        int hash = 31;
+        hash = hash * 17 + key.hashCode();
+        return hash % hashTable.length;
+    }
+
+    @Override
+    public Iterator<V> iterator() {
+        return new Iterator<V>() {
+            int counterArray = 0;
+            int valuesCounter = 0;
+            Iterator<Node<K, V>> subIterator = null;
+
+            @Override
+            public boolean hasNext() {
+                if (valuesCounter == size)
+                    return false;
+                if (subIterator == null || !subIterator.hasNext()) {
+                    if (moveToNextCell()) {
+                        subIterator = hashTable[counterArray].getNodes().iterator();
+                    } else {
+                        return false;
+                    }
+                }
+                return subIterator.hasNext();
+            }
+
+            private boolean moveToNextCell() {
+                counterArray++;
+                while (hashTable[counterArray] == null) {
+                    counterArray++;
+                }
+                return hashTable[counterArray] != null;
+            }
+
+            @Override
+            public V next() {
+                valuesCounter++;
+                return subIterator.next().getValue();
+            }
+        };
+    }
+
+    //create node with key and value = null
+    //get list node and add the node
+    //and increment the size
+    private boolean simpleAdd(int index, Node<K, V> newNode) {
+        hashTable[index] = new Node<>(null, null);
+        hashTable[index].getNodes().add(newNode);
+        size++;
+        return true;
+    }
+
+    private class Node<K, V> {
+        private List<Node<K, V>> nodes;
```

Comment:
Node should not contain list of nodes, it is semantically wrong. Same class should not represent FirstNode and Node